[ENH] Link scaling

juangpc opened the pull request:
This update solves some thing that have been not working properly for some time regarding the scaling controls. 
The scaling sliders worked reversed (left = higher amplitude) and sometimes the graph was glitchy.
This solves these issues. Internally the methods to process the callback for each scaling control update (double spinboxes or sliders) and are now independent. 
I've implemented a link (by default) between MAGs and GRADs channel amplitudes. Which can be set by pressing the SHIFT key. 
I've also done a bit of cleaning.

gabrielbmotta commented on libraries/disp/viewers/channelselectionview.cpp:
```diff
@@ -210,7 +210,7 @@ void ChannelSelectionView::setCurrentlyMappedFiffChannels(const QStringList &map
     m_pUi->m_listWidget_selectionGroups->clear();
 
```
I'm now seeing multiple 'All' groups in the channel selection

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -167,72 +155,185 @@ class DISPSHARED_EXPORT ScalingView : public AbstractView
+
+    //=============================================================================================================
+
+    void updateGUI();
```
add documentation to function

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -167,72 +155,185 @@ class DISPSHARED_EXPORT ScalingView : public AbstractView
+    //=========================================================================================================
+    /**
+     * Link the Magnetometers to the Gradiometers scale.
+     * @param [in] l Set the link between MAGs and GRADs active or inactive.
```
add line between function doc and parameter doc (see other functions as well)

gabrielbmotta commented on libraries/disp/viewers/channelselectionview.cpp:
```diff
@@ -410,7 +410,7 @@ void ChannelSelectionView::updateDataView()
             }
         }
-        SelectionSceneItem* test = static_cast<SelectionSceneItem*>(visibleItemList.first());
+        //SelectionSceneItem* test = static_cast<SelectionSceneItem*>(visibleItemList.first());
```
line if not needed

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.cpp:
```diff
@@ -0,0 +1,231 @@
+#include "scalecontrol.h"
+#include "ui_scalecontrol.h"
+
+//#include <QDebug>
```
remove includes if not needed

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,80 @@
+#include <QWidget>
+
+//class QLabel;
+//class QDoubleSpinBox;
```
remove includes if not needed

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,80 @@
+    Q_OBJECT
+
+public:
+    ScaleControl(const char* label);
```
Add documentation to functions

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,80 @@
+public:
+    ScaleControl(const char* label);
+    ScaleControl(const char* label, QWidget* parent);
+    ScaleControl(const char* label, QWidget* parent, double min, double max);
```
one parameter per line, per code conventions (see other functions as well)

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.cpp:
```diff
@@ -0,0 +1,231 @@
+#include "scalecontrol.h"
```
Missing mnecpp header

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,80 @@
+#ifndef SCALECONTROL_H
```
missing mnecpp header

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,80 @@
+    float                   m_fMaxSensitivityPoint;         /**< Max sensitivity point of the non-linear mapping fcn for the slider. */
+    float                   m_fMapYconstant;                /**< Y constant in the non-linear mapping curve for the slider. */
+    float                   m_fMapKconstant;                /**< K constant in the non-linear mapping curve for the slider. */
+    float                   m_bSliderInverted;              /**< State variable to store the inverted or not state of the slider. */
```
seems to be used throughout the code as a bool but is declared a float

gabrielbmotta reviewed (changes requested):
Great scaling!

gabrielbmotta commented on libraries/disp/viewers/channelselectionview.cpp:
```diff
@@ -455,6 +454,13 @@ void ChannelSelectionView::loadSettings()
+//    QList<QScreen*> screensList = QGuiApplication::screens();
+//    if(screensList.isEmpty())
+//    {
+//        move(QPoint(100,100));
```
remove commented code if no longer necessary

juangpc replied:
This was part of the code that solved a warning which we added back to trace some error. I'll add it back in.

gabrielbmotta commented on libraries/disp/viewers/channelselectionview.cpp:
```diff
@@ -589,7 +596,7 @@ bool ChannelSelectionView::loadSelectionGroups(QString path)
 
     //Create group 'All' and 'All EEG' manually (bcause this group depends on the loaded channels from the Info data file, not on the loaded selection file)
-    m_selectionGroupsMap["All"] = m_currentlyLoadedFiffChannels;
+    m_selectionGroupsMap["All"] =  m_currentlyLoadedFiffChannels;
```
remove extra space after =

juangpc replied:
done

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.cpp:
```diff
@@ -0,0 +1,258 @@
+/**
+ * @file     scalecontrol.cpp
+ * @author   Juan Garcia-Prieto <jgarciaprieto@mgh.harvard.edu>
+ * @since    0.1.0
```
version should be 0.1.9

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.cpp:
```diff
@@ -0,0 +1,258 @@
+void ScaleControl::initSpinBox()
+{
+    m_pUi->spinBox->setKeyboardTracking(false);
+//    m_pUi->spinBox->setSingleStep(0.01);
```
remove commented code if no longer necessary

juangpc replied:
OK

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+/**
+ * @file     scalecontrol.h
+ * @author   Juan Garcia-Prieto <jgarciaprieto@mgh.harvard.edu>
+ * @since    0.1.0
```
version should be 0.1.9

juangpc replied:
OKbis

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    //=========================================================================================================
+     /**
+     * getUI Retunr a pointer to the GUI of the ScaleControl.
+     * @return Pointer to the ScaleControlWidget.
```
add line between function doc and return doc.
remove function name from function doc
Retunr -> Return

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    //=========================================================================================================
+    /**
+     * Return the value of the control
+     * @return value of the control.
```
add line between function doc and return doc.

gabrielbmotta replied:
issue still present

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    //=========================================================================================================
+    /**
+     * Set the value of the maximum sensitivity point for the non-linear sensitivity curve of the sensor.
+     * @param [in] s Maximum sensitivity value
```
add line between param doc and return doc.

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    //=========================================================================================================
+    /**
+     * Set the sensitivity value of the non-linear mapping between the control slider and the control value.
+     * @param [in] s Sensitivity, between [0, 1]
```
add line between param doc and return doc.

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    //=========================================================================================================
+    /**
+     * @brief setRange
+     * @param min
```
add line between param doc and return doc.
document function and parameters

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    //=========================================================================================================
+    /**
+     * Set the number of decimals to show in the spinbox of the control.
+     * @param d
```
add line between param doc and return doc.
document function and parameters

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    //=========================================================================================================
+    /**
+     * Invert the effect of sliding the control slider to the right. Slider right decreases the control value.
+     * @param [in] inverted [True = Invert the slider effect].
```
add line between param doc and return doc.

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    //=========================================================================================================
+    /**
+     * Set the value of the control.
+     * @param [in] dScale value to set the control to.
```
add line between param doc and return doc.

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    //=========================================================================================================
+    /**
+     * Signal emited when the value of the control changes.
+     * @param dScale
```
add line between param doc and return doc.

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+     */
+    void valueChanged(double dScale);
+
+private:
```
add line between param doc and return doc in all the private functions.
some functions missing parameter documentation

gabrielbmotta commented on libraries/disp/viewers/scalingview.cpp:
```diff
@@ -143,31 +153,70 @@ float DISPLIB::getScalingValue(const QMap<qint32, float>& qMapChScaling,
 // DEFINE MEMBER METHODS
 //=============================================================================================================
 
+
```
remove extra lines

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -118,6 +105,7 @@ DISPSHARED_EXPORT float getScalingValue(const QMap<qint32, float>& qMapChScaling
                                         int iChannelKind,
                                         int iChannelUnit);
 
+
```
remove extra line

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -167,72 +155,198 @@ class DISPSHARED_EXPORT ScalingView : public AbstractView
+    /**
+     * Renders visible the control object that sets the link between Magnetometers and Gradiometers used in the controlView.
+     */
+
```
remove line between doc and function

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -167,72 +155,198 @@ class DISPSHARED_EXPORT ScalingView : public AbstractView
+     */
+    void scalingChanged(const QMap<qint32, float>& scalingMap);
+
+private:
```
private functions missing parameter documentation

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -167,72 +155,198 @@ class DISPSHARED_EXPORT ScalingView : public AbstractView
+    bool                                    m_bManagingLinkMagToGrad;       /**< Bool member mutex the link between MAGs and GRADs. */
+    QPointer<QCheckBox>                     m_pCheckBox;                    /**< Stores the conversion ratio between MAGs and GRADs. */
 };
+
```
remove extra line

gabrielbmotta reviewed (changes requested):
some more nitpicks

gabrielbmotta commented:
I've realized we've not been following our own [coding conventions](https://mne-cpp.github.io/pages/development/contr_style.html) all the way. We have not been adding the 'version' tag to functions. @LorenzE @juangpc what are your thoughts on this.

juangpc commented:
> I've realized we've not been following our own [coding conventions](https://mne-cpp.github.io/pages/development/contr_style.html) all the way. We have not been adding the 'version' tag to functions. @LorenzE @juangpc what are your thoughts on this.

I think we should delete it from the conventions... git will tell anybody interested when was each method added and modified. I don't find it very useful to be honest. To work correctly that should be done automatically by some kind of code analysis tool. 

> some more nitpicks

I think I've addressed all your 'picks.

gabrielbmotta commented:
> I think I've addressed all your 'picks.

There are still some that were left unaddressed. Can we have the function documentation in the same style/formatting/spacing as in the coding conventions page?  I know this is somewhat tedious but the goal is to have the documentation look uniform across the whole project. So from the webpage ( without the since tag) the functions should look something like this:
```
//=========================================================================================================
/**
 * Constructs a FileLoader object.
 *
 * @param[in]  iOpenMode            The open mode. 0=ReadWrite, 1=WriteOnly, 1=ReadOnly
 * @param[in]  bSkipEmptyLines      Whether to skip empty lines.
 * @param[in]  sCommentIdentifier   String to identify comments. Default is empty which
 *                                  results in comments being read as normal lines.
 */
FileLoader(int iOpenMode,
           bool bSkipEmptyLines,
           const QString& sCommentIdentifier = "");
```

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+
+    //=========================================================================================================
+    /**
+     * Memberfunction to update the constant values of the mapping functions between spinbox and the slider. These constant
```
Memberfunction -> Member function

Might be unnecessary to state this is a member function

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+
+    //=========================================================================================================
+    /**
+     * Memberfunction to addapt the sensitivity value to different range levels.
```
Memberfunction -> Member function

Might be unnecessary to state this is a member function

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -167,72 +155,200 @@ class DISPSHARED_EXPORT ScalingView : public AbstractView
-     * Emit this signal whenever the scaling sliders or spin boxes changed.
+     * Slot for processing the use of shortcut releases.
+     *
+     * @param [in] event An keyboard-press event to be processed.
```
use A instead on An

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -167,72 +155,200 @@ class DISPSHARED_EXPORT ScalingView : public AbstractView
+    /**
+     * Slot for processing the use of shortcut keypresses.
+     *
+     * @param [in] event An keyboard-release event to be processed.
```
use A instead on An

juangpc replied:
done

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+/**
+ * DECLARE CLASS ScaleControl
+ *
+ * @brief The ScaleControl class packs together a QLable, a DoubleSpinbox and a Slider with the necessary facilities for the
```
QLabel instead of QLable

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+
+    //=========================================================================================================
+    /**
+     * Constructs a ScaleControl object who's Qlabel elemennt will be set, and sets the parent of the underlyning
```
elemennt -> element
underlyning- >underlying

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+
+    //=========================================================================================================
+    /**
+     * Constructs a ScaleControl object who's Qlabel elemennt will be set, and sets the parent of the underlyning
```
elemennt -> element
underlyning- >underlying

juangpc replied:
💯

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+signals:
+    //=========================================================================================================
+    /**
+     * Signal emited when the value of the control changes.
```
emited -> emitted

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    /**
+     * Member method to set the range of the slider. Normally this is set to [1,1000] and there should be no reason
+     * to change it since the non-linear mapping function between the slider and the spinbox should take care of
+     * addapting the actual value of the control (the spinbox) with the values of the slider. Not even if there is
```
addapting -> adapting

juangpc replied:
👍🏻

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,271 @@
+    /**
+     * Member function to map the value of a slider poisition with a value for the control or the spinbox. This
+     * functions adapts the ranges of the control/spinbox to the ranges of the slider for consistency between ranges.
+     * @param [in] iIn Integer value representing the position fo the slider.
```
fo -> of

juangpc replied:
done

gabrielbmotta reviewed (changes requested):
yet another round of nitpicks

juangpc commented:
> > I think I've addressed all your 'picks.
> 
> There are still some that were left unaddressed. Can we have the function documentation in the same style/formatting/spacing as in the coding conventions page? I know this is somewhat tedious but the goal is to have the documentation look uniform across the whole project. So from the webpage ( without the since tag) the functions should look something like this:
> 
> ```
> //=========================================================================================================
> /**
>  * Constructs a FileLoader object.
>  *
>  * @param[in]  iOpenMode            The open mode. 0=ReadWrite, 1=WriteOnly, 1=ReadOnly
>  * @param[in]  bSkipEmptyLines      Whether to skip empty lines.
>  * @param[in]  sCommentIdentifier   String to identify comments. Default is empty which
>  *                                  results in comments being read as normal lines.
>  */
> FileLoader(int iOpenMode,
>            bool bSkipEmptyLines,
>            const QString& sCommentIdentifier = "");
> ```

Yes I'll do my best. This will trigger another pr, probably. But this is necessary, I agree.

juangpc commented:
We might be close to have this ready. Thanks @gabrielbmotta for the attention to review this. Great!

gabrielbmotta commented:
> > > I think I've addressed all your 'picks.
> > 
> > 
> > There are still some that were left unaddressed. Can we have the function documentation in the same style/formatting/spacing as in the coding conventions page? I know this is somewhat tedious but the goal is to have the documentation look uniform across the whole project. So from the webpage ( without the since tag) the functions should look something like this:
> > ```
> > //=========================================================================================================
> > /**
> >  * Constructs a FileLoader object.
> >  *
> >  * @param[in]  iOpenMode            The open mode. 0=ReadWrite, 1=WriteOnly, 1=ReadOnly
> >  * @param[in]  bSkipEmptyLines      Whether to skip empty lines.
> >  * @param[in]  sCommentIdentifier   String to identify comments. Default is empty which
> >  *                                  results in comments being read as normal lines.
> >  */
> > FileLoader(int iOpenMode,
> >            bool bSkipEmptyLines,
> >            const QString& sCommentIdentifier = "");
> > ```
> 
> Yes I'll do my best. This will trigger another pr, probably. But this is necessary, I agree.

I am referring to this PR Juan, not the whole project. I'm just referring to changing the formatting of these files to conform to the standards. Just this. I appreciate the other pr to fix the issues that got through in the past but I'm just trying to fix the formatting of this one.

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+    //=========================================================================================================
+    /**
+     * Set the number of decimals to show in the spinbox of the control.
+     * @param[in] d Number of decimals to show in the spinbox.
```
add line

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+    //=========================================================================================================
+    /**
+     * Set the value of the control.
+     *@param[in] dScale Value to set the control to.
```
add line

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+    //=========================================================================================================
+    /**
+     * Set the text label of the Control.
+     *@param[in] label New label text.
```
add line

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+     * and minimum levels of the control.
+     *
+     * @param[in] dIn Value of the spinbox or the control.
+     * @return Value to set the slider to, consistent with the range of the slider.
```
add line between param and return

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+     * ranges.
+     *
+     * @param[in] iIn Integer value representing the position of the slider.
+     * @return Double value to set the spinbox to.
```
add line between param and return

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+     * Memberfunction to addapt the sensitivity value to different range levels.
+     *
+     * @param[in] fSensitivity Sensitivity value
+     * @return Weighted sensitivity value
```
add line between param and return

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -167,72 +155,202 @@ class DISPSHARED_EXPORT ScalingView : public AbstractView
+    /**
+     * Callback to process a change in the MAGs scale spinbox.
+     *
+     * @param[in] dScale
```
add parameter documentation

gabrielbmotta commented on libraries/disp/viewers/scalingview.h:
```diff
@@ -167,72 +155,202 @@ class DISPSHARED_EXPORT ScalingView : public AbstractView
+     * @param[in] scalingMap this is a list of scaling values ordered by integer keys. These keys are resolved with
+     * the use of defined macros with names of channel-types, vendors, etc.
+     */
+    void scalingChanged(const QMap<qint32, float>& scalingMap);
```
one line per parameter

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+     *@param[in] iMin Minimum value for the slider.
+     *@param[in] iMax Maximum value for the slider.
+     */
+    void setSliderRange(int min, int max);
```
one line per parameter

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+     * @param min
+     * @param max
+     */
+    void setRange(double min, double max);
```
one line per parameter

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+     *@param[in] min      Min value of the control.
+     *@param[in] max      Max value of the control.
+     */
+     explicit ScaleControl(const char* label, QWidget* parent, double min, double max);
```
one line per parameter

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.h:
```diff
@@ -0,0 +1,280 @@
+     *@param[in] label    Text for the Qlabel
+     *@param[in] parent   Parent of widget
+     */
+    explicit ScaleControl(const char* label, QWidget* parent);
```
one line per parameter

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.cpp:
```diff
@@ -0,0 +1,255 @@
+
+//=============================================================================================================
+
+void ScaleControl::setRange(double min, double max)
```
one line per parameter

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.cpp:
```diff
@@ -0,0 +1,255 @@
+
+//=============================================================================================================
+
+ScaleControl::ScaleControl(const char* label, QWidget* parent, double min, double max)
```
one line per parameter

gabrielbmotta commented on libraries/disp/viewers/helpers/scalecontrol.cpp:
```diff
@@ -0,0 +1,255 @@
+
+//=============================================================================================================
+
+ScaleControl::ScaleControl(const char* label, QWidget* parent)
```
one line per parameter

juangpc commented:
Thanks. Let's see now.

juangpc commented:
Lets see.
